waffle_hub/hub/hub.py

```diff
@@ -1793,3 +1812,123 @@ def inner(callback: ExportCallback, result: ExportWaffleResult):
             callback.start()
 
         return result
+
+    def _hpo_hub_objective(self, **kwargs) -> float:
+        trial = kwargs.get("trial", None)
+        dataset = kwargs.get("dataset", None)
+        metric = kwargs.get("metric", None)
+        selected_params = kwargs.get("advance_params", None)
+
+        hub = Hub.new(
+            name=f"trial_{trial.number}",
+            root_dir=self.hub_dir / "hpo",
+            task=self.task,
+            model_type=self.model_type,
+            model_size=self.model_size,
+        )
+        train_result = hub.train(
+            dataset=dataset,
+            epochs=kwargs.get("epochs", None),
+            image_size=kwargs.get("image_size", None),
+            batch_size=kwargs.get("batch_size", None),
+            pretrained_model=kwargs.get("pretrained_model", None),
+            letter_box=kwargs.get("letter_box", None),
+            device=kwargs.get("device", 0),
+            workers=kwargs.get("workers", 2),
+            advance_params=selected_params,
+            hold=kwargs.get("hold", True),
+        )
+
+        for metric_value in train_result.eval_metrics:
+            if metric_value["tag"] == metric:
+                result = metric_value["value"]
+                break
+        return float(result)
+
+    def hpo(
+        self,
+        dataset: Union[Dataset, str],
+        metric: str,
+        search_space: Union[dict, str],
+        dataset_root_dir: Union[str, Path] = None,
+        sampler: Union[dict, str] = None,
+        pruner: Union[dict, str] = None,
+        direction: str = None,
+        n_trials: int = None,
+        device: str = 0,
+        workers: int = 2,
+        hold: bool = True,
+        **kwargs,
+    ) -> dict:
```

Comment:
hpo function is need error handing such as interrupt.

Reply:
OK!